crates/matrix-sdk/src/room/timeline/event_handler.rs

```diff
@@ -0,0 +1,460 @@
+        // `reaction_map`), it can still be present in the timeline items
+        // directly with the raw event timeline feature (not yet implemented).
+        did_update |= self.update_timeline_item(&redacts, "redaction", |item| item.to_redacted());
+
+        if !did_update {
+            // We will want to know this when debugging redaction issues.
+            debug!(redaction_key = ?self.flow.to_key(), %redacts, "redaction affected no event");
+        }
+    }
+
+    fn add(&mut self, item: NewEventTimelineItem) {
+        self.event_added = true;
+
+        let NewEventTimelineItem { content, reactions } = item;
+        let item = EventTimelineItem {
+            key: self.flow.to_key(),
+            event_id: None,
+            sender: self.meta.sender.to_owned(),
+            content,
+            reactions,
+            origin_server_ts: self.meta.origin_server_ts,
+            is_own: self.meta.is_own_event,
+            encryption_info: self.meta.encryption_info.clone(),
+            raw: self.meta.raw_event.clone(),
+        };
+
+        let item = Arc::new(TimelineItem::Event(item));
+        let mut lock = self.timeline.items.lock_mut();
+        match &self.flow {
+            Flow::Local { .. }
+            | Flow::Remote { position: EventPosition::End, txn_id: None, .. } => {
+                lock.push_cloned(item);
+            }
+            Flow::Remote { position: EventPosition::Start, txn_id: None, .. } => {
+                lock.insert_cloned(0, item);
+            }
+            Flow::Remote { txn_id: Some(txn_id), event_id, position } => {
+                if let Some((idx, _old_item)) = find_event(&lock, txn_id) {
+                    // TODO: Check whether anything is different about the old and new item?
+                    lock.set_cloned(idx, item);
+                } else {
+                    debug!(
+                        %txn_id, %event_id,
+                        "Received event with transaction ID, but didn't find matching timeline item"
+                    );
+
+                    match position {
+                        EventPosition::Start => lock.insert_cloned(0, item),
+                        EventPosition::End => lock.push_cloned(item),
+                    }
+                }
+            }
+        }
+    }
+
+    /// Returns whether an update happened
+    fn maybe_update_timeline_item(
+        &self,
+        event_id: &EventId,
+        action: &str,
+        update: impl FnOnce(&EventTimelineItem) -> Option<EventTimelineItem>,
+    ) -> bool {
+        // No point in trying to update items with relations when back-
+        // paginating, the event the relation applies to can't be processed yet.
+        if matches!(self.flow, Flow::Remote { position: EventPosition::Start, .. }) {
+            return false;
+        }
+
+        let mut lock = self.timeline.items.lock_mut();
+        if let Some((idx, item)) = find_event(&lock, event_id) {
```

Comment:
I wonder if that could become a bottle neck if we have scrolled back a room for a bit ... but I'll put that off as a future problem.